### DISCUSSION TITLE
Server model design using instances rather than static

### DISCUSSION THREAD

**Discussion by laultman** [StereoKit/StereoKit#680] [General]
@maluoi StereoKit is a singleton static environment, and it makes complete sense as it is expected to execute on exactly ONE device on which it is deployed. Holographic remoting breaks this single device paradigm. I have devoted many hours to tracing code and cleaning up resources in the C++ library in an attempt to return the code to its original state at launch on a disconnect event in the remoting service.
I am miles deep in the C++ trying to find every instance where variables are already initialized and therefore can't be reinitialized. This causes runtime errors as would be expected when C++ tries to allocate memory.
My question, given this deep instance problem, is it possible to create a version of StereoKit that is NOT static? If it were not static then the normal C# GC could handle the destruction of the all the objects.
Would you consider such a branch?

**Comment by maluoi** [comment]
Is it possible? I don't know, I suspect some of SK's dependencies might need to be re-thought as well. There would be big caveats related to threading. And it would likely wreak havoc on SK's API ergonomics. I'm not keen on making SK completely non-static.

I do want to make SK more resilient to restarts though. For this, my thought has been that each module should move their state into a struct, which can then be wiped at shutdown and set to default upon initialization. Easier to ensure everything gets proper attention when it's all in one place!

**Comment by laultman** [comment]
@maluoi I'm focused on the Windows implementation due to my need for HoloLens but I'm not uninterested in other implementations and I know eventually there will be other equipment I must support. That said, my remote rendering implementation needs a shared environment where multiple HoloLens devices are all working independently but sharing an orchestrated environment of data management. I have a state machine built into the orchestration that manages the interactivity across multiple devices. To accommodate the SK code and its one-to-one machine model, I have implemented a VM/Edge compute dedicated the instances.
I agree with the struct single-state. I think the current issue is that SK modules (C++) have no knowledge of state-of-the-machine at all. There is not code to register state or to return it to a given state. The SK C# has nothing it can call to "reset" state or to manage a clean restart.
My work in the C++/C# so far has only identified points in the C++ where memory is allocated and must be destroyed to restart (reinitialize). It suffers the same problem with no concept of state. To implement this will require a fairly sizable project from my point of view because I/my team don't know SK's C++ that well. We do know that we don't want to implement a new native project or really even fork this one. My hope it to make it work in a remote environment and be able to restart it cleanly, then to give you the code in hopes it will at least be adopted. You can then create the stateful context when it is appropriate in your roadmap.
As I sit here and write this, I am more excited about the possibility of a stateful service. That would work so very well if I could terminate and return. I have a requirement from a Navy customer to switch "apps" in training exercises. Dynamics 365 Guides can't do this in its current implementation. The Navy training requires stateful replay. So far to make this work I have implemented a Windows Service app with a coordinated collection of background services, one of which is an SK instance. Because SK is static it has to be implemented as a singleton service instance. The overhead of a spin-up of a new VM is too latent. To complicate things, Microsoft's OpenXR extensions events are not surfaced from the Windows Sockets layer, so I don't know the state of the network connection, its either ON or OFF. Because of this, I am forced to restart completely then on first frame go check the machine state that I maintain to reload the app state. Its workable but far from ideal.
In heavy Wifi environments where there is much contention for Wifi bandwidth, think tradeshows, the network doesn't work more than 6 to 10 feet because of packet loss and packet collisions. SK is unusable in these environments, it simply terminates when the OpenXR extensions finally time-out. Unity in remote is not any better as both rely on the network layer. As a reminder, the remoting is to gain access to heavy duty GPUs for complex rendering and to manage the latency of cloud services and large object models. Unfortunately, remoting is all or nothing. You are either remoting or you are local.

**Reply by maluoi** [reply]
Most modules in SK have both an initialize function, and a shutdown function. There's some modules that may not have those, but they typically pair with some other module that can be set up to call an initialize/shutdown on them.

In C#, `SK.Quit()` will cause `SK.Shutdown` to get called, (automatically when using `SK.Run`, you need to call `SK.Shutdown` yourself when directly using `SK.Step`) which shuts down all modules that have a shutdown function.

**Comment by laultman** [comment]
I mentioned previously that remoting is all or nothing. I hope I am wrong! I am experimenting with a hybrid model where I split the state-machine into a working pair where the remote is a slave rather than a master. This does have implications on GPU and my models, but I have other software in the pipeline to manage the polygon load.

**Comment by maluoi** [comment]
For reference, here's what I meant by wrapping module state up into a struct: https://github.com/StereoKit/StereoKit/commit/45c909c27dbf86db516827f0764ca7f1a8800fc8

It's really just keeping the local global state all packaged up in one spot. It also prevents use of global initializers for setting the initial state, which is kinda the core problem. This needs done for basically allll of the files, which will be a chunk of work. I'll probably need to sacrifice a day or two to get it all done, I guess at least it's the kind of code you can probably enjoy a movie with :D
